### ISSUE TITLE
vectorize sct2e, scencd, and et2utc

### ISSUE THREAD

**Issue by michaelaye** [AndrewAnnex/SpiceyPy#310]
This time, it's a floating point array, so I expected numpy to work, but it doesn't:

```python
arr = np.array([3.9138485783327938e+13, 3.9138485783955961e+13,
       3.9138485784583984e+13, 3.9138485785839891e+13,
       3.9138485786467852e+13])
scid = -64
spice.sct2e(scid, arr)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-77-aee45d30f174> in <module>
----> 1 spice.sct2e(scid, arr)

~/miniconda3/envs/py37/lib/python3.7/site-packages/spiceypy/spiceypy.py in with_errcheck(*args, **kwargs)
     75     def with_errcheck(*args, **kwargs):
     76         try:
---> 77             res = f(*args, **kwargs)
     78             checkForSpiceError(f)
     79             return res

~/miniconda3/envs/py37/lib/python3.7/site-packages/spiceypy/spiceypy.py in sct2e(sc, sclkdp)
  11076     """
  11077     sc = ctypes.c_int(sc)
> 11078     sclkdp = ctypes.c_double(sclkdp)
  11079     et = ctypes.c_double()
  11080     libspice.sct2e_c(sc, sclkdp, ctypes.byref(et))

TypeError: only size-1 arrays can be converted to Python scalars
```

**Comment by michaelaye** [comment]
Similar issue for et2utc:

```python
arr = np.array([597205999.6889952,
 597205999.6985781,
 597205999.708161,
 597205999.7273246,
 597205999.7369065])
spice.et2utc(arr, "ISOD", 6)

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-81-41248dd07bc6> in <module>
----> 1 spice.et2utc(arr, "ISOD", 6)

~/miniconda3/envs/py37/lib/python3.7/site-packages/spiceypy/spiceypy.py in with_errcheck(*args, **kwargs)
     75     def with_errcheck(*args, **kwargs):
     76         try:
---> 77             res = f(*args, **kwargs)
     78             checkForSpiceError(f)
     79             return res

~/miniconda3/envs/py37/lib/python3.7/site-packages/spiceypy/spiceypy.py in et2utc(et, formatStr, prec, lenout)
   5071     :rtype: str
   5072     """
-> 5073     et = ctypes.c_double(et)
   5074     prec = ctypes.c_int(prec)
   5075     lenout = ctypes.c_int(lenout)

TypeError: must be real number, not list
```

**Comment by AndrewAnnex** [comment]
this is not a numpy support issue, this is because these functions have not been “vectorized” to loop through iterable data types. This is easily fixable by just using for comprehension or a regular for loop. A PR that implements the “vectorization” would be a welcome contribution.

-Andrew Annex

**Comment by jessemapel** [comment]
You could use np.vectorize to make this and #309 easier. You can recursively call the function again with np.vectorize to avoid having to re-write anything:

```
import numpy as np
import spiceypy as spice

def new_bods2c(name):
    if hasattr(name, "__iter__") and not isinstance(name, str):
        return np.vectorize(new_bods2c, otypes=[int])(name).tolist()
    else:
        return spice.bods2c(name)
    
def new_et2utc(et, formatStr, prec, lenout=256):
    if hasattr(et, "__iter__"):
        return np.vectorize(new_et2utc, otypes=[str], excluded=('formatStr', 'prec', 'lenout'))(et, formatStr, prec, lenout).tolist()
    else:
        return spice.et2utc(et, formatStr, prec, lenout)
```

Note that for #309 and vectorizing over strings you have to check if the input is a string because they are iterable themselves.

Here's some sample outputs

```
print(new_bods2c('EARTH'))
print(new_bods2c(['EARTH', 'MARS']))
```

> 399
> [399, 499]

```
spice.furnsh('naif0012.tls')
print(new_et2utc(0, 'ISOC', 4))
print(new_et2utc([10**x for x in np.arange(10)], 'ISOC', 4))
```

> 2000-01-01T11:58:55.8161
> ['2000-01-01T11:58:56.8161', '2000-01-01T11:59:05.8161', '2000-01-01T12:00:35.8161', '2000-01-01T12:15:35.8161', '2000-01-01T14:45:35.8161', '2000-01-02T15:45:35.8160', '2000-01-13T01:45:35.8157', '2000-04-26T05:45:35.8145', '2003-03-03T21:45:35.8146', '2031-09-09T13:45:30.8175']

**Comment by michaelaye** [comment]
Andrew had a slice error check for each step of the loop. If there’s a reason for that, we can’t use vectorize.

**Comment by michaelaye** [comment]
SPICEerrorcheck sorry

**Comment by AndrewAnnex** [comment]
The error checks are necessary, there is existing behavior in ICY vectorized functions IIRC where the first exception reached should be raised, rather than running each function independently and having a mix of valid and invalid values in the return. [https://github.com/AndrewAnnex/SpiceyPy/blob/master/spiceypy/spiceypy.py#L13245](https://github.com/AndrewAnnex/SpiceyPy/blob/master/spiceypy/spiceypy.py#L13245) is an example code block for the vectorization methodology we landed on a few years ago.

**Comment by AndrewAnnex** [comment]
@michaelaye @jessemapel I merged in #309, this should make vectorizing these three functions a fairly easy thing to do now. @michaelaye I also added a unit test that uses a pandas series of strings that should hopefully work transparently. A PR that implements vectorization of these functions along with unit tests that use lists, numpy arrays, and pandas series/dataframes would be a very welcome addition.

**Comment by michaelaye** [comment]
i'm on it, got 2 out of 3 done...

**Comment by michaelaye** [comment]
but i don't understand what you mean by merged #309 ? That was an issue, not a PR?

**Comment by michaelaye** [comment]
i also add utc2et to the list of vectorization
